README.md
# laravel-varspay
> A Laravel Package for working with Varspay seamlessly


## Installation

[PHP](https://php.net) 5.4+ or [HHVM](http://hhvm.com) 3.3+, and [Composer](https://getcomposer.org) are required.

To get the latest version of Laravel Paystack, simply require it

```bash
composer require mujhtech/laravel-varspay
```

Or add the following line to the require block of your `composer.json` file.

```
"mujhtech/laravel-varspay": "1.0.*"
```

You'll then need to run `composer install` or `composer update` to download it and have the autoloader updated.


Once Laravel Varspay is installed, you need to register the service provider. Open up `config/app.php` and add the following to the `providers` key.

```php
'providers' => [
    ...
    Mujhtech\Varspay\VarspayServiceProvider::class,
    ...
]
```

> If you use **Laravel >= 5.5** you can skip this step and go to [**`configuration`**](https://github.com/unicodeveloper/laravel-paystack#configuration)

* `Unicodeveloper\Paystack\PaystackServiceProvider::class`

Also, register the Facade like so:

```php
'aliases' => [
    ...
    'Varspay' => Mujhtech\Varspay\Facades\Varspay::class,
    ...
]
```


## Configuration

You can publish the configuration file using this command:

```bash
php artisan vendor:publish --provider="Mujhtech\Varspay\VarspayServiceProvider"
```

A configuration-file named `varspay-tech.php` with some sensible defaults will be placed in your `config` directory:

```php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | API KEY
    |--------------------------------------------------------------------------
    |
    |
    */

    'apiKey' => getenv('VARSPAY_API_KEY'),

];
```


## Usage

Open your .env file and add your api key like so:

```php
VARSPAY_API_KEY=vp_live_xxxxxxxxxxx
```
*If you are using a hosting service like heroku, ensure to add the above details to your configuration variables.*


Let me explain the fluent methods this package provides a bit here.
```php
/**
     * Fetch your current balance
     * @param 
     * @return array
     */
Varspay::getBalance()

/**
 * Alternatively, use the helper.
 */
varspay()->getBalance();

/**
     * Fetch all information about yourself
     * @param
     * @return array
     */
Varspay::getMe();

/**
 * Alternatively, use the helper.
 */
varspay()->getMe();

/**
     * Fetch all alerts 
     * @param $acctNo
     * @return array
     */
Varspay::getAlerts($acctNo);

/**
 * Alternatively, use the helper.
 */
varspay()->getAlerts($acctNo);


/**
     * Retrieve Account Number
     * @param $email
     * @return array
     */
Varspay::retrieveAccountNumber($email);

/**
 * Alternatively, use the helper.
 */
varspay()->retrieveAccountNumber($email);


/**
     * Change transaction pin
     * @param $email, $pin
     * @return array
     */
Varspay::changePin($email, $pin);

/**
 * Alternatively, use the helper.
 */
varspay()->changePin($email, $pin);

/**
     * Change password
     * @param $email, $password
     * @return array
     */
Varspay::changePassword($email, $password);

/**
 * Alternatively, use the helper.
 */
varspay()->changePassword($email, $password);


/**
     * Fetch list of available banks
     * @param
     * @return array
     */
Varspay::bankList();

/**
 * Alternatively, use the helper.
 */
varspay()->bankList();


/**
     * Fetch information about bvn (Note: This cost 25 per call)
     * @param $bvn
     * @return array
     */
Varspay::resolveBvn($bvn);

/**
 * Alternatively, use the helper.
 */
varspay()->resolveBvn($bvn);


/**
     * Fetch information about an account
     * @param $acctNo, $bankCode
     * @return array
     */
Varspay::resolveAccount($acctNo, $bankCode);

/**
 * Alternatively, use the helper.
 */
varspay()->resolveAccount($acctNo, $bankCode);


/**
     * Create a virtual account
     * @param $acctName
     * @return array
     */
Varspay::createVirtualAccount($acctName);

/**
 * Alternatively, use the helper.
 */
varspay()->createVirtualAccount($acctName);


/**
     * Fetch all virtual accounts
     * @param 
     * @return array
     */
Varspay::getAllVirtualAccount();

/**
 * Alternatively, use the helper.
 */
varspay()->getAllVirtualAccount();


/**
     * Fetch all transaction on a specific virtual account
     * @param $acctNo
     * @return array
     */
Varspay::getVirtualAccountTransaction($acctNo);

/**
 * Alternatively, use the helper.
 */
varspay()->getVirtualAccountTransaction($acctNo);


/**
     * Single transfer to varspay account
     * @param $amount, $email
     * @return array
     */
Varspay::transferToVarspay($amount, $email);

/**
 * Alternatively, use the helper.
 */
varspay()->transferToVarspay($amount, $email);


/**
     * Single transfer to other bank
     * @param $amount, $bankCode, $description, $acctNo, $acctName
     * @return array
     */
Varspay::transferToOtherBank($amount, $bankCode, $description, $acctNo, $acctName);

/**
 * Alternatively, use the helper.
 */
varspay()->transferToOtherBank($amount, $bankCode, $description, $acctNo, $acctName);


/**
     * Bulk transfer to other bank
     * @param $bulk = [ array( "amount" => "100", "narration" => "Test Transfer", "accountname" => "OLUCHI AMARA", "bankname" => "ECOBANK", "accountnumber" => "8883116789", "bankcode"=> "000010"), array( "amount" => "100", "narration" => "Test Transfer", "accountname" => "OLUCHI AMARA", "bankname" => "ECOBANK", "accountnumber" => "8883116789", "bankcode"=> "000010") ]
     * @return array
     */
Varspay::transferToOtherBankBulk($bulk);

/**
 * Alternatively, use the helper.
 */
varspay()->transferToOtherBankBulk($bulk);



/**
     * Query single transfer transaction
     * @param subaccount code
     * @return array
     */
Varspay::getSingleTransferDetails($ref);

/**
 * Alternatively, use the helper.
 */
varspay()->getSingleTransferDetails($ref);


/**
     * Queries bulk transfer transaction
     * @param $ref code
     * @return array
     */
Varspay::getBulkTransferDetails($ref);

/**
 * Alternatively, use the helper.
 */
varspay()->getBulkTransferDetails($ref);

/**
     * Fetch airtime provider
     * @param 
     * @return array
     */
Varspay::airtimeProviders();

/**
 * Alternatively, use the helper.
 */
varspay()->airtimeProviders();

/**
     * Purchase an airtime
     * @param $amount, $phone, $provider code
     * @return array
     */
Varspay::airtimePurchase($amount, $phone, $provider);

/**
 * Alternatively, use the helper.
 */
varspay()->airtimePurchase($amount, $phone, $provider);

/**
     * Fetch airtime transaction queires
     * @param $ref code
     * @return array
     */
Varspay::airtimeQueris($ref);

/**
 * Alternatively, use the helper.
 */
varspay()->airtimeQueris($ref);

/**
     * List of airtim swap providers
     * 
     * @return array
     */
Varspay::airtimeSwapProviders();

/**
 * Alternatively, use the helper.
 */
varspay()->airtimeSwapProviders();

/**
     * create an airtimeswap transaction
     * @param $amount, $phone, $provider code
     * @return array
     */
Varspay::airtimeSwap($amount, $phone, $provider);

/**
 * Alternatively, use the helper.
 */
varspay()->airtimeSwap($amount, $phone, $provider);
```


## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.



CHANGELOG.md
# Changelog

All Notable changes to `laravel-varspay` will be documented in this file

## 2020-11-20
- Initial release

src/VarspayServiceProvider.php
<?php


/*
 * This file is part of the Laravel Paystack package.
 *
 * (c) Mujtech Mujeeb <mujeeb.muhideen@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Mujhtech\Varspay;

use Illuminate\Support\ServiceProvider;
use Mujhtech\Varspay\Console\VarspayInstallCommand;


class VarspayServiceProvider extends ServiceProvider
{

     /*
    * Indicates if loading of the provider is deferred.
    *
    * @var bool
    */


    /**
     * Register services.
     *
     * @return void
     */
    public function register()
    {
        //
        $this->app->bind('laravel-varspay', function ($app) {

            return new Varspay;

        });
    }

    /**
    * Publishes all the config file this package needs to function
    */

    public function boot()
    {
        $config = realpath(__DIR__.'/../config/varspay-tech.php');

        $this->publishes([
            $config => config_path('varspay-tech.php')
        ]);
        $this->registerCommands();
    }


    /**
    * Get the services provided by the provider
    * @return array
    */

    public function provides()
    {
        return ['laravel-varspay'];
    }


    private function registerCommands()
    {
        $this->commands([
            VarspayInstallCommand::class,
        ]);
    }

}

src/Console/VarspayInstallCommand.php
<?php

/*
 * This file is part of the Laravel Paystack package.
 *
 * (c) Mujtech Mujeeb <mujeeb.muhideen@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Mujhtech\Varspay\Console;

use Illuminate\Console\Command;

class VarspayInstallCommand extends Command {

    protected $signature = 'laravel-varspay:install '.
        '{--force : Overwrite existing views by default}'.
        '{--type= : Installation type, Available type: none, enhanced & full.}'.
        '{--only= : Install only specific part, Available parts: assets, config, translations, auth_views, basic_views, basic_routes & main_views. This option can not used with the with option.}'.
        '{--with=* : Install basic assets with specific parts, Available parts: auth_views, basic_views, basic_routes & main_views}'.
        '{--interactive : The installation will guide you through the process}';

    protected $description = 'Install all the required files for Varspay';


    protected $package_path = __DIR__.'/../../';

    

    /**
     * Execute the console command.
     *
     * @return void
     */
    public function handle()
    {
        if ($this->option('only')) {
            switch ($this->option('only')) {
            case 'config':
                $this->exportConfig();
                break;

            default:
                $this->error('Invalid option!');
                break;
            }

            return;
        }

        if ($this->option('type') == 'basic' || $this->option('type') != 'none' || ! $this->option('type')) {
            $this->exportConfig();
        }

        $this->info('Laravel Varspay Installation complete.');
    }


    /**
     * Install the config file.
     */
    protected function exportConfig()
    {
        if ($this->option('interactive')) {
            if (! $this->confirm('Install the package config file?')) {
                return;
            }
        }
        if (file_exists(config_path('varspay-tech.php')) && ! $this->option('force')) {
            if (! $this->confirm('The Laravel Varspay configuration file already exists. Do you want to replace it?')) {
                return;
            }
        }
        copy(
            $this->packagePath('config/varspay-tech.php'),
            config_path('varspay-tech.php')
        );

        $this->comment('Configuration Files installed successfully.');
    }

    /**
     * Get Package Path.
     */
    protected function packagePath($path)
    {
        return $this->package_path.$path;
    }

    /**
     * Get Protected.
     *
     * @return array
     */
    public function getProtected($var)
    {
        return $this->{$var};
    }


}

src/Varspay.php
<?php


/*
 * This file is part of the Laravel Paystack package.
 *
 * (c) Mujtech Mujeeb <mujeeb.muhideen@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Mujhtech\Varspay;
use Illuminate\Support\Facades\Config;

class Varspay {


    /**
	 * @var string Varspay.com API Endpoint
	 */

    private $endpoint = 'https://api.varspay.com/v1';

	/**
	 * @var string API Public Key
	 */
    private $apiKey;
    

    public function __construct() {
		$this->setApiKey();
    }
    

    /**
     * Set api key
	 * @param string $apiKey API Public Key
	 * @return 
	 */
	public function setApiKey() {
		$this->apiKey = Config::get('varspay-tech.apiKey');
    }


    /**
     * Fetch your current balance
     * @param 
     * @return array
     */


    public function getBalance(){

        $result = $this->request('user/balance', [], 'POST');

        return $result;

    }


    /**
     * Fetch all information about yourself
     * @param
     * @return array
     */


    public function getMe(){

        $result = $this->request('me', [], 'POST');

        return $result;

    }


    /**
     * Fetch all alerts 
     * @param $acctNo
     * @return array
     */


    public function getAlerts($acctNo){

        $result = $this->request('alerts?accountNumber='.$acctNo, [], 'GET');

        return $result;

    }


    /**
     * Retrieve Account Number
     * @param $email
     * @return array
     */


    public function retrieveAccountNumber($email){

        $result = $this->request('user/nuban', ['userEmail' => $email], 'POST');

        return $result;

    }


    /**
     * Change transaction pin
     * @param $email, $pin
     * @return array
     */


    public function changePin($email, $pin){

        $result = $this->request('user/pin', ['email' => $email, 'newPin' => $pin], 'POST');

        return $result;

    }

    /**
     * Change password
     * @param $email, $password
     * @return array
     */

    public function changePassword($email, $password){

        $result = $this->request('user/password', ['email' => $email, 'newPassword' => $password], 'POST');

        return $result;

    }


    /**
     * Fetch list of available banks
     * @param
     * @return array
     */

    public function bankList(){

        $result = $this->request('banks', [], 'GET');

        return $result;

    }


    /**
     * Fetch information about bvn (Note: This cost 25 per call)
     * @param $bvn
     * @return array
     */


    public function resolveBvn($bvn){

        $result = $this->request('resolve-bvn', ['bvn' => $bvn], 'POST');

        return $result;

    }


    /**
     * Fetch information about an account
     * @param $acctNo, $bankCode
     * @return array
     */


    public function resolveAccount($acctNo, $bankCode){

        $result = $this->request('resolve-account', [
            'accountnumber' => $acctNo,
            'bankcode' => $bankCode
        ], 'POST');

        return $result;

    }


    /**
     * Create a virtual account
     * @param $acctName
     * @return array
     */


    public function createVirtualAccount($acctName){

        $result = $this->request('createvirtualaccount', [
            'name' => $acctName
        ], 'POST');

        return $result;

    }

    /**
     * Fetch all virtual accounts
     * @param 
     * @return array
     */


    public function getAllVirtualAccount(){

        $result = $this->request('virtualaccount', [], 'POST');

        return $result;

    }

    /**
     * Fetch all transaction on a specific virtual account
     * @param $acctNo
     * @return array
     */


    public function getVirtualAccountTransaction($acctNo){

        $result = $this->request('virtualaccount/listtransactions', [
            'virtualaccount' => $acctNo
        ], 'POST');

        return $result;

    }

    /**
     * Single transfer to varspay account
     * @param $amount, $email
     * @return array
     */

    public function transferToVarspay($amount, $email){

        $result = $this->request('transfer/own', [
            'amount' => $amount,
            'email' => $email
        ], 'POST');

        return $result;

    }

    /**
     * Single transfer to other bank
     * @param $amount, $bankCode, $description, $acctNo, $acctName
     * @return array
     */

    public function transferToOtherBank($amount, $bankCode, $description, $acctNo, $acctName){

        $result = $this->request('transfer/other', [
            'amount' => $amount,
            'bankCode' => $bankCode,
            'description' => $description,
            'accountNumber' => $acctNo,
            'accountName' => $acctName
        ], 'POST');

        return $result;

    }

    /**
     * Bulk transfer to other bank
     * @param $bulk = [ array( "amount" => "100", "narration" => "Test Transfer", "accountname" => "OLUCHI AMARA", "bankname" => "ECOBANK", "accountnumber" => "8883116789", "bankcode"=> "000010"), array( "amount" => "100", "narration" => "Test Transfer", "accountname" => "OLUCHI AMARA", "bankname" => "ECOBANK", "accountnumber" => "8883116789", "bankcode"=> "000010") ]
     * @return array
     */

    public function transferToOtherBankBulk($bulk){

        $result = $this->request('transfer/other/bulk', [
            'bulkList' => $bulk
        ], 'POST');

        return $result;

    }

    /**
     * Query single transfer transaction
     * @param subaccount code
     * @return array
     */


    public function getSingleTransferDetails($ref){

        $result = $this->request('transfer/details/single?reference='.$ref, [], 'GET');

        return $result;

    }


    /**
     * Queries bulk transfer transaction
     * @param $ref code
     * @return array
     */

    public function getBulkTransferDetails($ref){

        $result = $this->request('transfer/details/bulk?batchReference='.$ref, [], 'GET');

        return $result;

    }


    /**
     * Fetch airtime provider
     * @param 
     * @return array
     */


    public function airtimeProviders(){

        $result = $this->request('airtime/providers', [], 'POST');

        return $result;

    }


    /**
     * Purchase an airtime
     * @param $amount, $phone, $provider code
     * @return array
     */


    public function airtimePurchase($amount, $phone, $provider){

        $result = $this->request('airtime/buy', [
            'amount' => $amount,
            'phoneNumber' => $phone,
            'providerShortCode' => $provider
        ], 'POST');

        return $result;

    }

    /**
     * Fetch airtime transaction queires
     * @param $ref code
     * @return array
     */

    public function airtimeQueris($ref){

        $result = $this->request('airtime/query', [
            'reference' => $ref
        ], 'POST');

        return $result;

    }

    /**
     * List of airtim swap providers
     * 
     * @return array
     */

    public function airtimeSwapProviders(){

        $result = $this->request('airtime/swap/providers', [], 'POST');

        return $result;

    }

    /**
     * create an airtimeswap transaction
     * @param $amount, $phone, $provider code
     * @return array
     */

    public function airtimeSwap($amount, $phone, $provider){

        $result = $this->request('airtime/swap', [
            'amount' => $amount,
            'phoneNumber' => $phone,
            'provider' => $provider
        ], 'POST');

        return $result;

    }
    



    /**
	 * @param string $method
	 * @param array $params
     * @param array $type
	 */
	private function request($method, $params = [], $type) {

		if (empty($this->apiKey) ) {
			return 'Error: API Key are not set.';
		}

		$url = $this->endpoint . '/' . $method;


        if($type == "POST"){

            $data = json_encode($params, 320);
        
            $curl = curl_init();

            curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => "",
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_SSL_VERIFYPEER => false,
            CURLOPT_SSL_VERIFYHOST => false,
            CURLOPT_CUSTOMREQUEST => "POST",
            CURLOPT_POSTFIELDS => $data,
            CURLOPT_HTTPHEADER => array(
                "Accept: application/json",
                "Content-Type: application/json",
                "Authorization: Bearer ".$this->apiKey
            ),
            ));
            
            $response = curl_exec($curl);
            $err = curl_error($curl);
            curl_close($curl);
            
            
            if ($err) {
                return "cURL Error #:" . $err;
            } else {
                return json_decode($response, true);
            }

        } elseif($type == "GET"){

            $curl = curl_init();

            curl_setopt_array($curl, array(
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => "",
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_SSL_VERIFYPEER => false,
            CURLOPT_SSL_VERIFYHOST => false,
            CURLOPT_CUSTOMREQUEST => "GET",
            CURLOPT_HTTPHEADER => array(
                "Accept: application/json",
                "Content-Type: application/json",
                "Authorization: Bearer ".$this->apiKey
            ),
            ));

            $response = curl_exec($curl);
            $err = curl_error($curl);
            curl_close($curl);
            
            
            if ($err) {
                return "cURL Error #:" . $err;
            } else {
                return json_decode($response, true);
            }

        } else {
            return 'Undefined Method Type';
        }
	}

}